### PR TITLE
Refactor SpaceDashboard Tab page object

### PR DESCRIPTION
**Pull request by jarifibrahim**
- Add Space dashboard Header
- Move Planner specific `ui` files to `ui/planner`

### CODE REVIEW

File: ee_tests/src/specs/page_objects/app/header.ts [hunk 1]
```diff
@@ -46,16 +46,19 @@ export class RecentItemsDropdown extends ui.Dropdown {
...
 
 export class Header extends ui.BaseElement {
+  mainNavBar: ui.BaseElement;
   profileDropdown = new ProfileDropdown(this.$('.pull-right.dropdown'));
   recentItemsDropdown = new RecentItemsDropdown(this.$$('.dropdown').get(0));
   // statusDropdown = new Dropdown(this.$$('.dropdown').get(0));
 
-  constructor(el: ElementFinder) {
-    super(el);
+  constructor(ele: ElementFinder = $('alm-app-header')) {
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/app/header.ts]: Please pass in the object instead of setting the default as that makes it clear when you are initialising header what element it is pointing to.

File: ee_tests/src/specs/page_objects/app/header.ts [hunk 2]
```diff
@@ -46,16 +46,19 @@ export class RecentItemsDropdown extends ui.Dropdown {
...
+    super(ele);
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/app/header.ts]: why is `el` renamed to `ele` with no apparent benefit ?

File: ee_tests/src/specs/page_objects/app/header.ts [hunk 3]
```diff
@@ -46,16 +46,19 @@ export class RecentItemsDropdown extends ui.Dropdown {
 
 
 export class Header extends ui.BaseElement {
+  mainNavBar: ui.BaseElement;
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/app/header.ts]: how about following the convention used everywhere 
```
mainNavBar = new ui.BaseElement(this.$('alm-app-header'));

File: ee_tests/src/specs/page_objects/planner/main.page.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { $, browser } from 'protractor';
+
+import { AppPage } from '../app.page';
+import { WorkItem } from './../../support/core';
+import * as planner from './../../ui/planner/';
+import * as support from './../../support';
+
+// this is what you see when you click on the Plan Tab button
+export class Planner extends AppPage {
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/planner/main.page.ts]: Call it `PlannerPage` since it is a page

File: ee_tests/src/specs/page_objects/space_dashboard.page.ts [hunk 5]
```diff
@@ -11,6 +11,11 @@ import { AppPage } from './app.page';
 import * as ui from '../ui';
 import { AddToSpaceDialog } from './space_dashboard/add_to_space_dialog';
 import { TextInput, Button } from '../ui';
+import { WorkItem } from './../support/core';
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/space_dashboard.page.ts]: Seems wrong .. why is WorkItem part of core?

File: ee_tests/src/specs/page_objects/space_dashboard.page.ts [hunk 6]
```diff
@@ -11,6 +11,11 @@ import { AppPage } from './app.page';
...
+import { debug } from './../support/core';
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/space_dashboard.page.ts]: since support is already imported, why not use support.debug() ?

File: ee_tests/src/specs/page_objects/space_dashboard.page.ts [hunk 7]
```diff
@@ -228,17 +152,19 @@ export class SpaceDashboardPage extends SpaceTabPage {
   wizardSidebar = element(by.xpath('.//*[contains(@class,\'wizard-pf-sidebar\')]'));
 
   spaceName: string;
-
+  header = new SpaceDashboardPageHeader();
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/space_dashboard.page.ts]: must be passed in the elementFinder so that it reads something like `header = new SpaceDashpardPageHeader(this.$('sub-element'))

File: ee_tests/src/specs/support/core.ts [hunk 8]
```diff
@@ -173,3 +173,10 @@ export function targetPlatform(): string {
   return 'fabric8-openshift';
 }
 
+type WorkItemType = 'task' | 'feature' | 'bug';
```
Comment (sthaha) [inline ee_tests/src/specs/support/core.ts]: this must be in the `ui/planner/index.ts`

File: ee_tests/src/specs/page_objects/app/header.ts (hunk 1 above)
Comment (jarifibrahim) [reply ee_tests/src/specs/page_objects/app/header.ts]: :+1:

File: ee_tests/src/specs/page_objects/app/header.ts (hunk 3 above)
Comment (jarifibrahim) [reply ee_tests/src/specs/page_objects/app/header.ts]: `SpaceDashboardPage` has a different `mainNavBar` and `LandingPage` has a different `mainNavBar`. That's why I've declared it here.

**Comment by jarifibrahim** [comment]
@sthaha I've made the requested changes but now there seems to be a circular dependency problem
```
[23:28:27] E/launcher - Error: TypeError: Class extends value undefined is not a function or null
    at Object.<anonymous> (/home/ijarif/Desktop/test-setup/fabric8-test/ee_tests/src/specs/page_objects/planner/main.page.ts:8:34)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/ijarif/Desktop/test-setup/fabric8-test/ee_tests/src/specs/page_objects/space_dashboard/space_dashboard_tab.ts:3:1)
    at Module._compile (module.js:570:32)
Closing report
[23:28:27] E/launcher - Process exited with error code 100
```

`PlannerPage` extends `SpaceDashboardTab`
`SpaceDashboardTab` contains `gotoPlanTab: PlannerPage` method which return a `PlannerPage` object.

**Comment by jarifibrahim** [comment]
@sthaha I propose the following
We use a common `Header` class for `Planner`, `Analytics` and `Space dashboard` pages. The common header would be property in all of these pages. Something like
```
Class CustomHeader extends Header {
  planTab = ...

  goToPlanTab() {
     ...  
   }
}

class PlannerPage extends AppPage {
    Header = CustomHeader(..)
}
```

Right now we extend pages from `SpaceTab` and all of the pages (plan, analysis, space dashboard) have nothing in common except the `header`.
We've already had the discussion regarding the `spaceTabPage` class but it would be better if you could explain it once again :)

File: ee_tests/src/specs/page_objects/planner/planner.page.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    support.debug(' ... check if Planner page is Ready');
+    await super.ready();
+    await this.workItemList.ready();
+    await this.quickAdd.ready();
+    support.debug(' ... check if Planner page is Ready - OK');
+  }
+
+  async createWorkItem(item: planner.WorkItem) {
+    this.debug('create item', JSON.stringify(item));
+    await this.quickAdd.createWorkItem(item);
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/planner/planner.page.ts]: why is this an async method?

File: ee_tests/src/specs/ui/space_dashboard/index.ts [hunk 10]
```diff
@@ -0,0 +1,2 @@
+export { AddToSpaceDialog } from './add_to_space_dialog';
+export { SpaceDashboardHeader } from './space_dashboard_header';
```
Comment (sthaha) [inline ee_tests/src/specs/ui/space_dashboard/index.ts]: should be spacedashboard/header.ts  just like app/header.ts

File: ee_tests/src/specs/ui/space_dashboard/space_dashboard_header.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  createTab = new ui.Clickable(
+    this.$('.persistent-secondary').element(by.cssContainingText('li', 'Create')),
+    'Create');
+
+  constructor(el: ElementFinder, spaceName: string, name: string = 'Space Dashboard page header') {
+    super(el, name);
+    this.spaceName = spaceName;
+  }
+  async ready() {
+    support.debug(' ... check if SpaceDashboardPageHeader is Ready');
```
Comment (sthaha) [inline ee_tests/src/specs/ui/space_dashboard/space_dashboard_header.ts]: Have you considered using this.debug which will print the className as well?

there is also a better way to doing `ready` which logs better messages when `ready` starts and ends see ui/base.element.ts - ready and run

File: ee_tests/src/specs/ui/planner/workitem-quickadd.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  constructor(el: ElementFinder, name = 'Work Item Quick Add') {
+    super(el, name);
+  }
+
+  async ready() {
+    await super.ready();
+    await this.untilClickable();
+  }
+
+  async createWorkItem({ title, description = '', type = 'feature' }: WorkItem) {
```
Comment (sthaha) [inline ee_tests/src/specs/ui/planner/workitem-quickadd.ts]: how about `addWorkItem` since the name of the component is `QuickAdd`

File: ee_tests/src/specs/page_objects/app.page.ts [hunk 13]
```diff
@@ -13,10 +13,10 @@ export abstract class AppPage extends BasePage {
   /**
    * Extend this class, to describe Application Page(after logging in)
    *
-   * @param {url} elementFinder ElementFinder that you want to extend
+   * @param {url} string URL where the extended page resides
    */
-  constructor() {
-    super();
+  constructor(url?: string) {
```
Comment (sthaha) [inline ee_tests/src/specs/page_objects/app.page.ts]: Please note the comments about url? in base.page.ts. URL set to '' is a valid one.

Why was this change needed by the way? Can this be made into a separate commit but part of the same PR?

File: ee_tests/src/specs/workshop-import-to-space.spec.ts [hunk 14]
```diff
@@ -4,9 +4,9 @@ import {
   MainDashboardPage,
   SpaceDashboardPage,
   CleanupUserEnvPage,
-  AddToSpaceDialog,
   PageOpenMode
 } from './page_objects';
+import { AddToSpaceDialog } from './ui/space_dashboard/';
```
Comment (sthaha) [inline ee_tests/src/specs/workshop-import-to-space.spec.ts]: why / at the end? how about export {AddToSpaceDialog} in space_dashboard.page.ts?

File: ee_tests/src/specs/page_objects/planner/planner.page.ts (hunk 9 above)
Comment (jarifibrahim) [reply ee_tests/src/specs/page_objects/planner/planner.page.ts]: Because I can `await` only inside an `async` method

File: ee_tests/src/specs/ui/space_dashboard/space_dashboard_header.ts (hunk 11 above)
Comment (jarifibrahim) [reply ee_tests/src/specs/ui/space_dashboard/space_dashboard_header.ts]: @sthaha I decided to do it similar to https://github.com/fabric8io/fabric8-test/blob/master/ee_tests/src/specs/page_objects/app/header.ts#L14

Do you think we should do it
```
async ready() {
   this.run('... check if SpaceDashboardPageHeader is Ready', async () => {
    ...
    ...
  })
}
``` 
this way?

File: ee_tests/src/specs/page_objects/app.page.ts (hunk 13 above)
Comment (jarifibrahim) [reply ee_tests/src/specs/page_objects/app.page.ts]: The parent class `base.page's` constructor accepts a url/undefined value.
Using `url?: string` would ensure user passes only a string and not a boolean/number.
Also, if the user doesn't pass anything `url` would be undefined, as we expect it to be.

**Comment by jarifibrahim** [comment]
Closing this PR since we no longer need it.

**Comment by ldimaggi** [comment]
We can close the pull request. I've added all the required code to https://github.com/fabric8-ui/fabric8-planner/pull/2417
